hardware_integration/src/peak_ethernet_gateway_plugin.cpp

```diff
@@ -0,0 +1,388 @@
+#include "isobus/hardware_integration/peak_ethernet_gateway_plugin.hpp"
+
+#include "isobus/isobus/can_stack_logger.hpp"
+
+#include <cstring>
+
+#ifndef _MSC_VER
+#include <arpa/inet.h>
+#include <netinet/tcp.h>
+#include <unistd.h>
+#endif
+
+namespace isobus
+{
+#ifdef __GNUC__
+#define PACK(__Declaration__) __Declaration__ __attribute__((__packed__))
+#endif
+
+#ifdef _MSC_VER
+#define PACK(__Declaration__) __pragma(pack(push, 1)) __Declaration__ __pragma(pack(pop))
+#endif
+
+	/// @brief This is the schema of the message payload the gateway uses
+	/// @returns Packed version of _CAN2IP, as defined by Peak
+	typedef PACK(struct _CAN2IP {
+		std::uint16_t sz;
+		std::uint16_t type;
+		std::uint64_t tag;
+		std::uint64_t timestamp;
+		std::uint8_t channel;
+		std::uint8_t dlc;
+		std::uint16_t flag;
+		std::uint32_t id;
+		std::uint8_t data[8];
+	}) S_CAN2IP;
+
+	PeakEthernetGatewayPlugin::PeakEthernetGatewayPlugin(std::string targetIPAddress,
+	                                                     std::uint16_t txPort,
+	                                                     std::uint16_t rxPort,
+	                                                     bool useUDP) :
+	  ipAddress(targetIPAddress),
+	  sendPort(txPort),
+	  receivePort(rxPort),
+	  udp(useUDP)
+	{
+	}
+
+	PeakEthernetGatewayPlugin::~PeakEthernetGatewayPlugin()
+	{
+		PeakEthernetGatewayPlugin::close();
+#ifdef WIN32
+		WSACleanup();
+#endif
+	}
+
+	bool PeakEthernetGatewayPlugin::get_is_valid() const
+	{
+		return isOpen;
+	}
+
+	void PeakEthernetGatewayPlugin::close()
+	{
+		if (0 != rxSocket)
+		{
+#ifdef WIN32
+			closesocket(rxSocket);
+#else
+			::close(rxSocket);
+#endif
+		}
+		if (0 != txSocket)
+		{
+#ifdef WIN32
+			closesocket(txSocket);
+#else
+			::close(txSocket);
+#endif
+		}
+	}
+
+	void PeakEthernetGatewayPlugin::open()
+	{
+		if (!isOpen)
+		{
+			bool rxOpened = open_rx_socket();
+
+			if (0 != rxSocket)
+			{
+				if (rxOpened)
+				{
+					bool txOpened = open_tx_socket();
+
+					if (txOpened)
+					{
+						rxBuffer.reset(new std::uint8_t[RX_BUFFER_SIZE_BYTES]());
```

Comment:
Hahah mixing a little old C stuff with C++'s smart pointers, it does it's job I guess 😆